osu.Game/Skinning/Skin.cs

```diff
@@ -293,20 +304,71 @@ private void applyMigration(SkinLayoutInfo layout, GlobalSkinnableContainers tar
                         c.Type.Name == nameof(DefaultComboCounter) ||
                         c.Type.Name == nameof(ArgonComboCounter)).ToArray();
 
-                    layout.Update(null, globalHUDComponents.Except(comboCounters).ToArray());
+                    hudLayout.Update(null, globalHUDComponents.Except(comboCounters).ToArray());
 
                     resources.RealmAccess.Run(r =>
                     {
                         foreach (var ruleset in r.All<RulesetInfo>())
                         {
-                            layout.Update(ruleset, layout.TryGetDrawableInfo(ruleset, out var rulesetHUDComponents)
+                            hudLayout.Update(ruleset, hudLayout.TryGetDrawableInfo(ruleset, out var rulesetHUDComponents)
                                 ? rulesetHUDComponents.Concat(comboCounters).ToArray()
                                 : comboCounters);
                         }
                     });
 
                     break;
                 }
+
+                case 2:
+                {
+                    // Legacy health bar is moved from the global HUD components target into per-ruleset.
+                    // On osu!mania, the health bar is moved from the HUD to Playfield target for stage-based positioning.
+                    if (resources == null)
+                        break;
+
+                    var hudLayout = layoutInfos.GetValueOrDefault(GlobalSkinnableContainers.MainHUDComponents);
+                    var playfieldLayout = layoutInfos.GetValueOrDefault(GlobalSkinnableContainers.Playfield);
+
+                    if (hudLayout == null || !hudLayout.TryGetDrawableInfo(null, out var globalHUDComponents))
+                        globalHUDComponents = Array.Empty<SerialisedDrawableInfo>();
+
+                    var legacyHealthBars = globalHUDComponents.Where(h => h.Type.Name == nameof(LegacyHealthDisplay)).ToArray();
+                    hudLayout?.Update(null, globalHUDComponents.Except(legacyHealthBars).ToArray());
+
+                    resources.RealmAccess.Run(r =>
+                    {
+                        foreach (var ruleset in r.All<RulesetInfo>())
+                        {
+                            if (ruleset.ShortName == @"mania")
+                            {
+                                // should avoid adding legacy health bar to non-legacy skins (unless explicitly added by user).
+                                if (!legacyHealthBars.Any())
+                                    break;
```

Comment:
This implies that if the skin being migrated here has configured the `Playfield` target but never configured the non-ruleset `HUD` target, then the health bar will not be added.

It's a hard case to handle here because we don't really have a nice way of telling whether a skin is based on `LegacySkin`, nor do I want to use that as a condition anywhere anyway. I think it's fine to let those kind of skins manually add their health bar by resetting the target or otherwise.

Reply:
I've read this comment five times and I still can't understand what you're trying to say. Can you elaborate? Doesn't the HUD target read fall back to an empty array if it fails? Can't you use that?

Reply:
> Doesn't the HUD target read fall back to an empty array if it fails?

What does "fails" mean here.

Let me attempt to explain how the migration logic works in a case-by-case scenario:
 - A skin with no configuration
   - `HUD` and `Playfield` are all null, so the lookups will get the default components from `LegacySkin` and `ManiaLegacySkinTransformer` if it's a legacy skin, we're gucci.
 - A skin with `HUD` configuration but no `Playfield` configuration
   - `HUD` is not null and `Playfield` is null
   - If the `HUD` configuration has a legacy health bar, remove it and store it to `legacyHealthBars`.
   - Don't do anything for `Playfield`, it's already null, the health bar will be looked up from `ManiaLegacySkinTransformer` if it's a legacy skin.
 - A skin with `HUD` configuration and `Playfield` configuration
   - `HUD` and `Playfield` are both not null
   - If the `HUD` configuration has a legacy health bar, remove it and store it to `legacyHealthBars`.
   - If `legacyHealthBars` is not empty, add a legacy mania health bar to `Playfield`.
     - The condition used here is good enough to ensure if the configured skin is not legacy then no legacy health bar will be added to mania, by checking whether the skin used to have a legacy health bar or not.
 - A skin with no `HUD` configuration but a `Playfield` configuration
   - `HUD` is null, `Playfield` is not null
   - Don't do anything for `HUD`, it's already null
   - `legacyHealthBars` is empty because `HUD` is null
   - Don't do anything for `Playfield` since `legacyHealthBars` is empty
     - This is the part touched on in this conversation. This is incorrect because if the skin is legacy then there should be a health bar added to `Playfield` by this migration logic, but I'm intentionally ignoring it for simplicity. It should be a rare case and one that can be fixed by the user.

Reply:
Might as well write test cases rather than state them word by word.

Reply:
Test cases written within an introduced testing system `TestSceneSkinMigration`.